src/Agent.Listener/Configuration/CredentialProvider.cs

```diff
@@ -312,4 +319,109 @@ public override void EnsureCredential(IHostContext context, CommandSettings comm
             CredentialData.Data[Constants.Agent.CommandLine.Args.Password] = command.GetPassword();
         }
     }
+
+
+    public sealed class DeviceCodeCredential : CredentialProvider
+    {
+        public DeviceCodeCredential() : base(Constants.Configuration.DeviceCode) { }
+
+        public override VssCredentials GetVssCredentials(IHostContext context)
+        {
+            ArgUtil.NotNull(context, nameof(context));
+            Tracing trace = context.GetTrace(nameof(DeviceCodeCredential));
+            trace.Info(nameof(GetVssCredentials));
+
+            CredentialData.Data.TryGetValue(Constants.Agent.CommandLine.Args.TenantId, out string tenantId);
+            ArgUtil.NotNullOrEmpty(tenantId, nameof(tenantId));
+            trace.Info("tenant id retrieved: {0} chars", tenantId.Length);
+
+            CredentialData.Data.TryGetValue(Constants.Agent.CommandLine.Args.ClientId, out string clientId);
+            ArgUtil.NotNullOrEmpty(clientId, nameof(clientId));
+            trace.Info("client id retrieved: {0} chars", clientId.Length);
+
+            var app = PublicClientApplicationBuilder.Create(clientId).WithTenantId(tenantId).Build();
+
+            var authResult = AcquireATokenFromCacheOrDeviceCodeFlowAsync(context, app, new string[] { "499b84ac-1321-427f-aa17-267ca6975798/user_impersonation" }).GetAwaiter().GetResult(); ;
```

Comment:
Maybe add some explaining comment for this id?

Reply:
Fixed by internal variable with short description.
This is Azure Devops resource id which is described here - https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/manage-personal-access-tokens-via-api?view=azure-devops